### DISCUSSION TITLE
Episode 171: Thinking Parallel & C++ Forward Progress

### DISCUSSION THREAD

**Discussion by codereport** [codereport/adsp2#63] [Episode]
Episode Link: https://adspthepodcast.com/2024/03/01/Episode-171.html

Feel free to write comments or ask questions 🙂

**Comment by Brian-M-J** [comment]
Speaking of parallelism, a really interesting project in the parallel programming space is the [Higher-Order Virtual Machine](https://github.com/HigherOrderCO/HVM):
> No need to deal with the complexity of concurrent programming: locks, mutexes, atomics... **any** work that can be done in parallel **will** be done in parallel.

Source - [HigherOrderCO company website](https://higherorderco.com/)

Achieving perfect automatic parallelization sounds like magic to me.

They also claim that they need no garbage collector (unlike Haskell's GHC):

> Since HVM2 is based on Interaction Combinators, which are fully linear, there is no global “garbage collection” pass required. By IC evaluation semantics alone, data is granularly allocated as needed, and freed as soon as they become unreachable. [...] When HVM2 completes evaluation [...], the memory should be left with just the final result of the program, and no remaining garbage or byproduct. No further work is needed.

Source - [HVM2: A Parallel Evaluator For Interaction Combinators](https://paper.higherorderco.com/)

To show off, they have this [example](https://github.com/HigherOrderCO/HVM/blob/main/POST.md#parallel-bitonic-sort):
> Implementing a Bitonic Sort in C-like languages can be challenging, as it involves coordination of shared memory accesses. [...] Unlike the CUDA counterpart, this version of the algorithm is extremely high-level, relying on millions of small allocations, tree rotations and recursion. It isn't even possible to express it that way in any existing GPU-targeting language. Bend is capable of compiling that algorithm, as is, to in more than 32 thousand CUDA threads, achieving near-ideal speedup

They have 2 posts on HackerNews where they explain more: [[1]](https://news.ycombinator.com/item?id=30219452) [[2]](https://news.ycombinator.com/item?id=35336113)
